src/platform/plugins/shared/discover/public/application/main/discover_main_route.test.tsx

```diff
@@ -77,120 +145,150 @@ describe('DiscoverMainRoute', () => {
-      expect(findTestSubject(component, 'noDataViewsPrompt').length).toBe(1);
-    });
-  });
+    setupComponent({ hasESData: true, hasUserDataView: false });
 
-  // skipped because this is the case that never ever should happen, it happened once and was fixed in
-  // https://github.com/elastic/kibana/pull/137824
-  test.skip('renders no data page when hasESData=false & hasUserDataView=true', async () => {
-    const component = mountComponent(false, true);
+    await waitForLoad();
 
-    await waitFor(() => {
-      component.update();
-      expect(findTestSubject(component, 'kbnNoDataPage').length).toBe(1);
-    });
+    expect(screen.getByTestId('noDataViewsPrompt')).toBeVisible();
   });
 
   test('renders LoadingIndicator while customizations are loading', async () => {
     let resolveService = (_: DiscoverCustomizationService) => {};
     mockCustomizationService = new Promise((resolve) => {
       resolveService = resolve;
     });
-    const component = mountComponent(true, true);
-    await waitFor(() => {
-      component.update();
-      expect(component.find(DiscoverMainApp).exists()).toBe(false);
-    });
+    setupComponent({ hasESData: true, hasUserDataView: true });
+
+    expect(screen.getByLabelText('Loading')).toBeInTheDocument();
+
     resolveService(createCustomizationService());
-    await waitFor(() => {
-      component.setProps({}).update();
-      expect(component.find(DiscoverMainApp).exists()).toBe(true);
-    });
+    await waitForLoad();
+
+    expect(screen.queryByLabelText('Loading')).not.toBeInTheDocument();
   });
 
   test('renders LoadingIndicator while root profile is loading', async () => {
     mockRootProfileState = { rootProfileLoading: true };
-    const component = mountComponent(true, true);
-    await waitFor(() => {
-      component.update();
-      expect(component.find(DiscoverMainApp).exists()).toBe(false);
-    });
-    mockRootProfileState = defaultRootProfileState;
-    await waitFor(() => {
-      component.setProps({}).update();
-      expect(component.find(DiscoverMainApp).exists()).toBe(true);
+
+    setupComponent({ hasESData: true, hasUserDataView: true });
+
+    expect(screen.getByLabelText('Loading')).toBeInTheDocument();
+  });
+
+  describe.each([
+    { hasChanged: false, id: undefined },
+    { hasChanged: true, id: undefined },
+    { hasChanged: false, id: '1234' },
+  ])('when there are no unsaved changes', ({ hasChanged, id }) => {
+    it('should call the default action', () => {
+      // Given
+      const tabMock = getTabRuntimeStateMock({
+        stateContainer$: new BehaviorSubject<DiscoverStateContainer | undefined>({
+          savedSearchState: {
+            getHasChanged$: () => new BehaviorSubject(hasChanged),
+            getId: () => id,
+          },
+          dataState: {
+            cancel: jest.fn(),
+          },
+          actions: {
+            stopSyncing: jest.fn(),
+          },
+        } as unknown as DiscoverStateContainer),
```

Comment:
I'm not a fan of the partial mock + cast approach since it makes tracking down test usages harder when refactoring, and would rather we use real instances. Would it work to instead mock `useStateManagers` and return real instances using `getDiscoverStateMock`? Another alternative would be to extract the functionality into a separate hook to test in isolation and pass an real instance as a param.